File: 1016. Phone Bills (25).cpp
#include <cstdio>
#include <cstring>
#include <string>
#include <algorithm>
#include <vector>
#include <map>

using namespace std;

struct Record {
    int dd, hh, mm, time;//dd * 24 * 60 + hh * 60 + mm
    bool state;//true:on-line false:off-line
    string name;
    Record():state(false){}
    Record(char *name, int dd, int hh, int mm, bool state)
        :dd(dd), hh(hh), mm(mm), time(dd * 24 * 60 + hh * 60 + mm),
        state(state), name(name){}
};

bool cmp(const Record &a, const Record &b){
    return a.name == b.name ? a.time < b.time : a.name < b.name;
}

double calRecond(Record r, int *rate){
    double res = r.dd * 60 * rate[24] + r.mm * rate[r.hh];
    for(int i = 0; i < r.hh; i++) res += 60 * rate[i];
    return res / 100;
}

int main(){
    int rate[25] = {0};
    for(int i = 0; i < 24; i++){
        scanf("%d", &rate[i]);
        rate[24] += rate[i];
    }
    int n, month, dd, hh, mm;
    char name[25], state[10];
    vector<Record> records;
    vector<Record>::iterator vit;
    scanf("%d", &n);
    for(int i = 0; i < n; i++){
        scanf("%s%d:%d:%d:%d%s", name, &month, &dd, &hh, &mm, state);
        Record record(name, dd, hh, mm, strcmp(state, "on-line") == 0 ? true : false);
        records.push_back(record);
    }
    sort(records.begin(), records.end(), cmp);
    map<string, vector<Record>> customer;
    map<string, vector<Record>>::iterator mit;
    for(int i = 1; i < n; i++){
        if(records[i - 1].name == records[i].name){
            if(records[i - 1].state && !records[i].state){
                customer[records[i].name].push_back(records[i - 1]);
                customer[records[i].name].push_back(records[i]);
            }
        }
    }
    for(mit = customer.begin(); mit != customer.end(); mit++){
        printf("%s %02d\n", mit->first.c_str(), month);
        vector<Record> rec = mit->second;
        double charge = 0;
        for(int i = 1; i < rec.size(); i += 2){
            double temp = calRecond(rec[i], rate) - calRecond(rec[i - 1], rate);
            charge += temp;
            printf("%02d:%02d:%02d ", rec[i - 1].dd, rec[i - 1].hh, rec[i - 1].mm);
            printf("%02d:%02d:%02d ", rec[i].dd, rec[i].hh, rec[i].mm);
            printf("%d $%.2f\n", rec[i].time - rec[i - 1].time, temp);
        }
        printf("Total amount: $%.2f\n", charge);
    }
    return 0;
}

File: 1016. Phone Bills (25).py
#calculate 0 to time cost
def calRecond(time, rate):
    res =  float(time[1] * 60.0 * rate[24] + time[3] * rate[time[2]]);
    for i in range(time[2]):
        res += 60 * rate[i];
    return res / 100

#input rate
rate = map(int, raw_input().strip().split())
rate.append(sum(rate))
#input n
n = int(raw_input().strip())
#input recond [name mm:dd:hh:mm onORoff]
reconds = []
for i in range(n):
    cin = map(str, raw_input().strip().split())
    time = map(int, cin[1].split(':'))
    recond = [cin[0], time, cin[2], time[1] * 24 * 60 + time[2] * 60 + time[3]]
    reconds.append(recond)
reconds.sort(key = lambda rec:(rec[0], rec[3]))
#choose match recond
customers = []
name = ""
for i in range(1, n):
    pre = reconds[i - 1]
    now = reconds[i]
    if pre[0] == now[0]:
        if pre[2] == "on-line" and now[2] == "off-line":
            if name == "" or name != pre[0]:
                customers.append([])
                name = pre[0]
            customers[len(customers) - 1].append(pre)
            customers[len(customers) - 1].append(now)
#output
for customer in customers:
    print "%s %02d" % (customer[0][0], customer[0][1][0])
    charge = 0
    for i in range(1, len(customer), 2):
        pre = customer[i - 1]
        now = customer[i]
        temp = calRecond(now[1], rate) - calRecond(pre[1], rate)
        charge += temp
        print "%02d:%02d:%02d" % (pre[1][1], pre[1][2], pre[1][3]),
        print "%02d:%02d:%02d" % (now[1][1], now[1][2], now[1][3]),
        print "%d $%.2f" % (now[3] - pre[3], temp)
    print "Total amount: $%.2f" % charge